packages/neo-one-monitor/src/BrowserMonitor.js

```diff
@@ -30,30 +55,193 @@ class BrowserGauge extends BaseMetric implements Gauge {
+  gauges: Array<BaseMetric>,
+  histograms: Array<BaseMetric>,
+  summaries: Array<BaseMetric>,
+|};
+
+class BrowserMetricsFactory implements MetricsFactoryCollect {
+  _metrics: MetricCollection;
+
+  constructor() {
+    this._metrics = {
+      counters: [],
+      gauges: [],
+      histograms: [],
+      summaries: [],
+    };
+  }
+
+  collect(): MetricCollection {
+    const metrics = JSON.parse(JSON.stringify(this._metrics));
+    for (const metricType of Object.keys(this._metrics)) {
+      this._metrics[metricType].map(metric => metric.reset());
+    }
+
+    return metrics;
+  }
+
   createCounter(options: MetricConstruct): Counter {
-    return new BrowserCounter(options);
+    const counter = new BrowserCounter(options);
+    this._metrics.counters.push(counter);
+
+    return counter;
   }
 
   createGauge(options: MetricConstruct): Gauge {
-    return new BrowserGauge(options);
+    const gauge = new BrowserGauge(options);
+    this._metrics.gauges.push(gauge);
+
+    return gauge;
   }
 
   createHistogram(options: MetricConstruct): Histogram {
-    return new BrowserHistogram(options);
+    const histogram = new BrowserHistogram(options);
+    this._metrics.histograms.push(histogram);
+
+    return histogram;
   }
 
   createSummary(options: MetricConstruct): Summary {
-    return new BrowserSummary(options);
+    const summary = new BrowserSummary(options);
+    this._metrics.summaries.push(summary);
+
+    return summary;
+  }
+}
+
+interface LoggerCollect extends Logger {
+  collect(): Array<LoggerLogOptions>;
+}
+
+export class BrowserLogger implements LoggerCollect {
+  _logs: Array<LoggerLogOptions>;
+
+  constructor() {
+    this._logs = [];
+  }
+
+  log(options: LoggerLogOptions): void {
+    this._logs.push(options);
+  }
+
+  collect(): Array<LoggerLogOptions> {
+    const logs = this._logs;
+    this._logs = [];
+
+    return logs;
+  }
+
+  close(callback: () => void): void {
+    callback();
+  }
+}
+
+export class Reporter {
+  _logger: BrowserLogger;
+  _monitor: BrowserMonitor;
+  _timer: number;
+  _endpoint: string;
+  _backLogs: Array<LoggerLogOptions>;
+  _backMetrics: MetricCollection;
+  _shutdownFunc: () => void;
+
+  constructor({
+    logger,
+    monitor,
+    timer,
+    endpoint,
+  }: {|
+    logger: BrowserLogger,
+    monitor: BrowserMonitor,
+    timer: number,
+    endpoint: string,
+  |}) {
+    this._logger = logger;
+    this._monitor = monitor;
+    this._timer = timer;
+    this._endpoint = endpoint;
+    this._backLogs = [];
+    this._backMetrics = {
+      counters: [],
+      gauges: [],
+      histograms: [],
+      summaries: [],
+    };
+
+    const subscription = interval(this._timer).subscribe(async () => {
+      let logs = this._logger.collect();
+      const metrics = this._monitor.collect();
+
+      if (this._backLogs.length > 0) {
+        logs = logs.concat(this._backLogs);
+        this._backLogs = [];
+
+        for (const metricType of Object.keys(metrics)) {
```

Comment:
This is a potentially expensive loop (`O(n^2)`) if there are a lot of metrics. Instead, you should refactor this so that you can just do a lookup to see if `_backMetrics` has the `name` you're looking for. I.e. `_backMetrics` should be something like an object from type => name => metric